### ISSUE TITLE
Analysis mode implementation

### ISSUE THREAD

**Issue by akmohtashami** [cms-dev/cms#605]
Regarding the implementation of analysis mode in CMS, do you think it is necessary?It needs quite some changes if we are going to allow users to submit in that phase.
What about cloning the contest instead? We could provide a method for cloning tasks and contests that would:
1- make all test-cases public
2- remove any submission limit
3- disable tokens(as there won't be any need for them and so the user would be able to see which submissions have been tokened during the contest)

Also if analysis mode is going to be implemented, what do you think about adding a boolean field to submissions  separating official/unofficial submissions. Unofficial submissions won't be send to RWS and won't be used in scoring.

What do you think?

**Comment by myungwoo** [comment]
What is the analysis mode you think? Could you please describe it more specifically?

**Comment by qnighy** [comment]
We (Japanese OI) currently do analysis by creating a backup of the contest. One of its problems is that exporting and importing contests need some time. (We have tighter schedule than IOI.)

I prefer "analysis mode" approach to "cloning" approach for at least one reason: the former seems to be faster than the latter. (Of course it depends on detail and implementation of these approaches.)

**Comment by akmohtashami** [comment]
@myungwoo Analysis mode is a phase after the contest finishes in which, contestants are able to see full detailed scores for all of their submissions. They are sometimes provided with testdata as well  in order to let them check the testdata and appeal if necessary. Also, it'd be nice to let them submit through CMS. (A similar explanation is given in #604)
 
The approach @qnighy is a fine approach and feels more secure since no change occurs in the main contest. The same was done in our national OI. 
However currently there are two problems. First is the one discussed by him. It takes quite some time if the testdata are large.
The other one is that last time I checked, due to updates to cmsDumpImporter, it is impossible to export and then import the contest again since it tries to import users as well which are existing and so an error is raised. (Actually I think this should be fixed regardless. I will see if I can send a PR to fix it).
Regarding the speed of this method, I think we can instead provide a cloning script(or even better a button in the admin interface) which doesn't re-import the files but uses the existing files digests(I am not completely sure that this is possible yet, but I looked through the code and it seemed fine). However I think the submissions can be actually cloned since they are usually low-sized and it shouldn't take long to clone them.

**Comment by qnighy** [comment]
If I remember correctly, the dumps might include compiled binaries of submissions. They might be also large (like >=1M per submission), since they are statically linked.

**Comment by akmohtashami** [comment]
I see. However I don't think we need to use the dumping system at all. We can just introduce task cloning which doesn't import any files and uses the already existing files instead. I thought about it a little more and I think it'd be fine if we'd do the same with the submissions too.

**Comment by stefano-maggiolo** [comment]
Yes, the usual dumping & reimporting has too many flaws. The approach of cloning, keeping reference to the existing files is interesting, and shouldn't take too much time. As for myself, what I've usually seen in big contests is to just dump the contest, and let contestant use the original one to see results and submit new solutions.

My opinion is that cloning is easier to implement (and actually might be useful in other situation?!), but as a solution for analysis mode is not too elegant. As I said before, my idea is to add another phase where there is full feedback and is pretty liberal with submissions.

Previously I thought about computing the score only using submission within the time of the contest, but actually I like Amir's idea to add a boolean that states whether it was submitted in the contest or in the analysis mode, as it handles better (I think) the case where the contest time changes.

**Comment by lucach** [comment]
A brief discussion which points to interesting resources has already happened in #103 (and related issues).

A few months ago I started working on the analysis mode with filtering based on the time of the submission (see lucach@04ca705 and lucach@b650c59). The approach does not seem too invasive, but the boolean flag on the submission appears to be simpler and safer.

**Comment by myungwoo** [comment]
I also agree with boolean field of the submission.

I want to make discussion about removing submission limit.
There are a few reasons why submission limit is necessary.
For example, to limit the submission loads, not just for restricting number of contestant's submission.
To keep submission loads not huge, submission limit is also necessary in analysis mode.

How do you think about it?

**Comment by gollux** [comment]
Is this still relevant?
